### PR TITLE
feat(ui/rest): User confirmation before updating records with duplicate attachments

### PR DESCRIPTION
**Pull request by smrutis1**
* Added a confirm dialog box to confirm with user before updating the record(release, project, component) which containes duplicate   attachment.
* Same validation is done at the REST side as well, if the client tries to upload duplicate attachment to the record(release, project, component), then it will give an error response with warning message. In case client wants to add duplicate attachment, then client needs to send additional flag with as a request parameter(`?allowDuplicateAttachment=true`) to achieve the same.

Signed-off-by: Smruti Sahoo <smruti.sahoo@siemens.com>

closes: #533

### CODE REVIEW

**Comment by nutanv1contr** [comment]
Two Issue:
1.Could you please format the text which is displayed as warning message.
2.Due to JS call failure for the function update release is not working.

**Comment by smrutis1** [comment]
@nutanv1contr Thanks for pointing this out, I have rectified it.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java [hunk 1]
```diff
@@ -186,4 +187,10 @@ private synchronized void makeConnector() throws TException {
         }
         return new Resources<>(attachmentResources);
     }
+
+    public boolean isDuplicateAttachment(Set<Attachment> attachments) {
+        boolean duplicateSha1 = attachments.parallelStream().collect(Collectors.groupingBy(Attachment::getSha1)).size() < attachments.size();
```
Comment (maxhbr) [inline rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java]: Maybe extract function into 
```
boolean checkForDuplicatesOn(Collection<Attachment> attachments, Function<Attachment,Object> extractor)
```

Comment (maxhbr) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/attachment/Sw360AttachmentService.java]: Might be easier to read as:
```suggestion
        boolean duplicateSha1 = attachments.stream().map(Attachment::getSha1).distinct().size() < attachments.size();
```

File: rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/integration/ComponentTest.java [hunk 2]
```diff
@@ -143,8 +150,10 @@ public void should_update_component_valid() throws IOException, TException {
         Map<String, String> body = new HashMap<>();
         body.put("name", updatedComponentName);
         body.put("invalid_property", "abcde123");
+        String url = "http://localhost:" + port + "/api/components/" + componentId;
+        UriComponentsBuilder builder = UriComponentsBuilder.fromUriString(url).queryParam("allowDuplicateAttachment", "true");
```
Comment (imaykay) [inline rest/resource-server/src/test/java/org/eclipse/sw360/rest/resourceserver/integration/ComponentTest.java]: Maybe add another test, where duplicates are not allowed and you try to update with a duplicate.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java [hunk 3]
```diff
@@ -136,11 +140,17 @@ public ResponseEntity searchByExternalIds(@RequestParam MultiValueMap<String, St
     @RequestMapping(value = RELEASES_URL + "/{id}", method = RequestMethod.PATCH)
     public ResponseEntity<Resource<Release>> patchComponent(
             @PathVariable("id") String id,
-            @RequestBody Release updateRelease) throws TException {
+            @RequestBody Release updateRelease,
```
[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java]: @imaykay Thanks for the nice suggestion with explanation. I would like to go with path variable as well in this case.

[the remark this answers is not in the document]
Comment (smrutis1) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java]: @lepokle @maxhbr  Any other opinion on this or shall we go ahead with using path variable.

**Comment by maxhbr** [comment]
Why is there the possibility to force duplicate attachments? In my opinion having this is not a good idea. E.g. if someone automatically updates projects via REST, his implementation will fail when a project has duplicate attachments and he has not thought about enforcing the duplicate attachment flag.

File: rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxhbr) [reply rest/resource-server/src/main/java/org/eclipse/sw360/rest/resourceserver/release/ReleaseController.java]: I have no strong opinion here

**Comment by maxhbr** [comment]
Ok, I understand that there might already be projects which contain duplicate attachments. But for them it might be a better choice to have something like: if the project previously contained duplicates, just skip the check?

**Comment by smrutis1** [comment]
On my opinion, it will be good to inform the user that project/component/release contains duplicate attachments which the user might have added accidentally. So that he can choose to delete the  duplicates.

**Comment by maxhbr** [comment]
> On my opinion, it will be good to inform the user that project/component/release contains duplicate attachments which the user might have added accidentally. So that he can choose to delete the duplicates.

Yes, this is good. But not for REST, where no intelligent person is involved in the interaction.

**Comment by mcjaeger** [comment]
what situation do we have
1. Attachments are duplicated already at components|projects|releases
2. there should not be any duplicates
3. However, if an attachment exists and it is a duplicate, it is maybe even locked, because of projet usage
4. Forcing no duplicates on save creates a blocking preventing to update a release|project|component

Solutions (so far)
1. duplicate check on Thrift level: consistent, but creates the block on existing records, as described above
2. do a check only in the UI (and no in REST and Thrift) showing a warning, but allowing the user to save intentionally: solves the case to some extent, but not on REST. Also REST and UI are inconsistent
3. do a check only in the UI (and no in REST and Thrift) showing a warning, but allowing the user to save intentionally and block it in REST, however, optionally provide flag to override duplicate blocking: works, but complicated.

Frankly I do not have a good idea what to do. Solution 1 is blocking us on our instance. Maybe a solution could be something completely different, like writing a DB migration script to eliminate duplicates on our production instance.

**Comment by mcjaeger** [comment]
(from discussion) let's have this now that before editin step, backend shall understand if already duplicate attachment exists or if not, in case a duplicate attachment exists, backend does not refuse update otherwise the duplicate attachment is added in the editing step and the backend should refuse.

By this way, it is consistent with either portal or REST API usage.
